katsdpcontroller/product_config.py

```diff
@@ -31,63 +82,1196 @@ class _Sensor(ABC):
+        self.excise = excise
+        self.archive = archive
+        self.n_servers = n_servers
+
+    @property
+    def baseline_correlation_products(self) -> BaselineCorrelationProductsStreamBase:
+        return cast(BaselineCorrelationProductsStreamBase, self.src_streams[0])
+
+    @property
+    def n_channels(self) -> int:
+        rng = self.output_channels
+        return (rng[1] - rng[0]) // self.continuum_factor
+
+    @property
+    def n_spectral_channels(self) -> int:
+        """Number of CBF channels that are in the output, before continuum averaging."""
+        return self.output_channels[1] - self.output_channels[0]
+
+    @property
+    def antennas(self) -> Sequence[str]:
+        return self.baseline_correlation_products.antennas
+
+    @property
+    def n_antennas(self) -> int:
+        return self.baseline_correlation_products.n_antennas
+
+    @property
+    def n_pols(self) -> int:
+        return 2          # TODO: get from config?
+
+    @property
+    def n_baselines(self) -> int:
+        a = self.n_antennas
+        return a * (a + 1) // 2 * self.n_pols**2
+
+    @property
+    def n_vis(self) -> int:
+        return self.n_baselines * self.n_channels
+
+    @property
+    def size(self) -> int:
+        """Size of each frame in bytes."""
+        # complex64 for vis, uint8 for weights and flags, float32 for weights_channel
+        return self.n_vis * BYTES_PER_VFW + self.n_channels * 4
+
+    @property
+    def flag_size(self):
+        """Size of the flags in each frame, in bytes."""
+        return self.n_vis * BYTES_PER_FLAG
+
+    def net_bandwidth(self, ratio: float = 1.05, overhead: int = 128) -> float:
+        return bandwidth(self.size, self.int_time, ratio, overhead)
+
+    def flag_bandwidth(self, ratio: float = 1.05, overhead: int = 128) -> float:
+        return bandwidth(self.flag_size, self.int_time, ratio, overhead)
+
+    @classmethod
+    def from_config(cls,
+                    options: Options,
+                    name: str,
+                    config: Mapping[str, Any],
+                    src_streams: Sequence['Stream'],
+                    sensors: Mapping[str, Any]) -> 'VisStream':
+        output_channels = config.get('output_channels')
+        if output_channels is not None:
+            output_channels = tuple(output_channels)
+        return cls(
+            name, src_streams,
+            int_time=config['output_int_time'],
+            output_channels=output_channels,
+            continuum_factor=config['continuum_factor'],
+            excise=config.get('excise', True),
+            archive=config['archive'],
+            n_servers=4 if not options.develop else 2
+        )
+
+    def compatible(self, other: 'VisStream') -> bool:
+        """Determine whether the configurations are mostly the same.
+
+        Specifically, they must be the same other than the vlaues of
+        ``continuum_factor`` and ``archive``.
+        """
+        return (
```

Comment:
wouldn't it be more robust to make copies of self and other, remove continuum_factor and archive and then compare them directly ? Saves updating this section if you add new things.

Reply:
Yes, I wasn't really happy with this, but thought that trying to compare two streams for equality would still require listing all the attributes. But this is Python, so I've solved it with some reflection.